### DISCUSSION TITLE
Data files and format

### DISCUSSION THREAD

**Discussion by fitneva** [jspsych/jsPsych#884] [Q&A] [answered]
Hi, 
In the examples we've seen so far in the workshop, the data pop up on the screen. What command is used to actually store participant data? Is there a choice of output file format? Sorry if I've missed that.

Related to that, I assume the data are recorded in a similar way as what we've seen on the screen. And that's far from the one trial per line that is usually needed for statistical analyses. Any advice on how to approach this data pre-processing step? Is there any code to use as a starting point?

Thanks!
Stanka

**Comment by conbainbridge** [comment, marked as answer]
There's more information on data saving with jsPsych here: https://www.jspsych.org/core_library/jspsych-data/

- The code you would use to save the data will depend on how/where you want to save it (to a server using http methods potentially using an API, to [MTurk](https://www.jspsych.org/core_library/jspsych-turk/), etc.).
- Data output types in jsPsych: JSON and CSV
-You could in theory use code to parse out parts of the data, but you likely don't need to as JSON can make this simple enough after the fact and it can be risky to drop data that may be valuable, such as the time elapsed. I don't have experience pre-processing from the front-end code as I've always kept all data, but if other people have thoughts/experience and have code they'd want to share, please do!

**Reply by becky-gilbert** [reply]
Just to add to Connie's response: there's an example of saving the data with PHP and MySQL in the jsPsych docs [here](https://www.jspsych.org/overview/data/#storing-data-permanently-in-a-mysql-database). But, as Connie says, the way that you save the data will depend on where you want it to be saved and what server-side tools you're using. 

For instance, if you plan to use [Pushkin](https://languagelearninglab.gitbook.io/pushkin/) then you'll hear about how to do this in the next workshop session. 
Or if you're not running the task on a server, then you could save the data to a local file (see docs [here](https://www.jspsych.org/core_library/jspsych-data/#localsave)). Or if you're using [JATOS](http://www.jatos.org/) to host your experiment, then you would use the `jatos.submitResultData()` function that they provide. 

And as far as data saving formats, you might also be interested in [this thread](https://github.com/jspsych/jsPsych/discussions/885) that mentions JSON vs CSV, with links to the relevant jsPsych documentation for each of these. If you do save the data in JSON format, then you can convert it to the standard one-row-per-trial table format using some kind of JSON-to-CSV converter. There are lots of options in different programming languages. I do this using an R package called [jsonlite](https://cran.r-project.org/web/packages/jsonlite/vignettes/json-aaquickstart.html).

Is this the information you were looking for? Please let us know if you have follow up questions.
